lib/models/cql_calculator.js

```diff
@@ -0,0 +1,643 @@
+      }
+    }
+
+    const observations = measure.get('observations');
+    const observationDefs = [];
+    if (observations) {
+      observations.forEach((obs) => {
+        generatedELMJSON = this.constructor.generateELMJSONFunction(
+          obs.function_name,
+          obs.parameter
+        );
+        // Save the name of the generated define statement, so we can check
+        // its result later in the CQL calculation process. These added
+        // define statements are called 'EcqmeFunction_' followed by the
+        // name of the function - see the 'generateELMJSONFunction' function.
+        observationDefs.push(`EcqmeFunction_${obs.function_name}`);
+        // Check to see if the gneratedELMJSON function is already in the definitions
+        // Added a check to support old ELM representation and new Array representation.
+        elm[mainLibraryIndex].library.statements.def.push(generatedELMJSON);
+      });
+    }
+
+    // Set all value set versions to 'undefined' so the execution engine does not grab the specified version in the ELM
+    elm.forEach((elmLibrary) => {
+      elmLibrary.library.valueSets.def.forEach((valueSetParam) => {
+        const valueSet = valueSetParam;
+        if (valueSet.version != null) {
+          valueSet.version = undefined;
+        }
+      });
+    });
+
+    // Grab the correct version of value sets to pass into the exectuion engine.
+    const measureValueSets =
+        this.valueSetsForCodeService(
+          measure.get('value_sets'),
+          measure.get('hqmf_set_id'),
+        );
+
+      // Calculate results for each CQL statement
+    const resultsRaw =
+        this.constructor.executeEngine(
+          elm, patientSource, measureValueSets,
+          measure.get('main_cql_library'), mainLibraryVersion, params,
+        );
+
+    Object.keys(resultsRaw.patientResults).forEach((patientId) => {
+      const result = CqmModels.IndividualResult();
+
+      // Parse CQL statement results into population values
+      const [populationResults, episodeResults] =
+            Array.from(this.createPopulationValues(
+              measure, resultsRaw,
+              patientId, observationDefs
+            ));
+        // All of the following until the final result set patient_id may no longer be needed.
+      if (populationResults != null) {
+        result.set(populationResults);
+        if (episodeResults != null) {
+          // In episode of care based measures, episodeResults contains the population results
+          // for EACH episode.
+          result.set({ episode_results: episodeResults });
+        }
+        result.patient = patientId;
+        result.measure = measure._id;
+        resultsByPatient[patientId] = result;
+      }
+    });
+    // } catch (error) {
+    //   throw Error(error);
+    // }
+
+    return resultsByPatient;
+  }
+
+  static executeEngine(elm, patientSource, valueSets, libraryName, version, parameters = {}) {
```

Comment:
Either pass in the new `executionDateTime` that I mentioned above into this function and pass it into `executor.exec(patientSource, executionDateTime)` or set `executionDateTime` in this function. This function could probably be refactored. It is currently a copy of a function that Bonnie calls in the CQL execution engine repo called `executeSimpleELM`, now that its being brought into this repo I think it would make sense to move around some of the logic. Thoughts?